rabbitmq/rabbitmq-management
# Support IdentityServer4 OAuth2 token provider

## Proposed Changes

Please consider to add support for scenarios where Rabbit Management backend handle both UAA and IdentityServer4 (IS4) OAuth2 token providers. 

Proposed solution uses OIDC (OpenId-Connect library) client to connect both Rabbit Management UI javascript backend and IdentityServer. In this case OIDC was configured to handle signin/signout redirections, store and management of returned JWT token in browser memory and automatically handle to refresh token when it's about to expire soon.

First attempt to integrate Management UI with IdentityServer assumed to obligatory use modified _rabbitmq-auth-backend-oauth2_ pluggin but now it is optional. I will provide manual for both scenarios. 
This implementation add few more options to _rabbitmq_management_ secrion, but with respect to UAA implementation there are shared parameters for both providers.

Below example config has **arrows** for new parameters used in IS4 case:

```
{rabbitmq_management, [
       {enable_oauth2, true},
       {uaa_client_id, "RabbitMq.ManagementUI"},
       {uaa_location, "http://localhost:50000"},
 --->  {oauth2_scopes, "read write add delete"}, <---
 --->  {oauth2_implementation, identityserver},  <---  %% uaa or identityserver
]},
```
_oauth2_scopes_ parameter was configured as string
_oauth2_implementation_ parameter was configured as enum [uaa , identityserver] with defeault _uaa_ value when its not configured

Options like: _enable_oauth2_, _uaa_client_id_ and _uaa_location_ are shared by UAA and IdentityServer implementations. Their meainng in both cases are the same.

Why do I need _oauth2_scopes_ parameter? 
Analyzed UAA implementation shows that Singular in rpFrame.js has hard coded "openid" scope with whom is requesting to UAA server. I have considered to make Singular use _oauth2_scopes_ as well, but I am not entirely sure about Singular origin, why he's code looks like obfuscated in Rabbit repo and any other but potentailly consequences regardled to UAA.

 ```
rpFrame.js

t=S("openid",h,g);
function S(t,n,r){return a.uaaLocation+"/oauth/authorize?response_type=token&scope="+encodeURIComponent(t)+"&client_id="+a.clientId+"&prompt=none&redirect_uri=.....
```

I found that when IdentityServer is requested without any scope assigned to IS4's ApiResource (ApiResource = resource_server_id) it returns token where aud (audience) section hasn't have included information about configured _resource_server_id_ - in another words it has no information that this token is valid for configured _resource_server_id_.

`eyJhbGciOiJSUzI1NiIsImtpZCI6Ik9YTndkX1RMZGJJY09leDZiTExTdFEiLCJ0eXAiOiJKV1QifQ.eyJuYmYiOjE1NzU4MTE4NDIsImV4cCI6MTYwNzM2ODc2OCwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo1MDAwMCIsImF1ZCI6IkdpdEh1Yi5SYWJiaXRNcS5NYW5hZ2VtZW50V2Vic2l0ZS5Mb2NhbCIsIm5vbmNlIjoiZmI1NWFjODRjYzEzNDMxZDg5YTFiMGY0YTAyNTE4NTQiLCJpYXQiOjE1NzU4MTE4NDIsImF0X2hhc2giOiJnYnY3RVB4WGJ4d0Q1bU9lVFAtLVFRIiwic19oYXNoIjoiV2dtb05aT3VoUWNsQWJNaW10YlllUSIsInNpZCI6IlBHVnNCM2lDTnFlTkhuYlJ2RGVKb0EiLCJzdWIiOiIxMTEiLCJhdXRoX3RpbWUiOjE1NzU4MTE4MDUsImlkcCI6ImxvY2FsIiwiYW1yIjpbInB3ZCJdfQ.pYBEUUe3ViAY-LDVwB5QP1fW0VFuxH017i16zvLjjDaZKJZVFLxu0EEc0SeNUD-kaifPPzB9d0KlCspptewsoGpFRbPfCFrTUOVl2G18QvW2ENUKMIxtGB_rMQLfeAEn9lYARwr1-W4i-ej_ZC2M00BlfUXiPyEYDGDRziE2z3kYmHODDlhZcrmLsu_BF0GQcaQzUj4HA9-ntw5OwkM9F1md9OhYtAzjIjTqIgkX8h9cw7u_cp_wAIlaQikxNeDQP3KO-IrwW39bkX6z0i3uk8oLOOC9h3bcd01BLMnzfor_UlN8VCZkEDTXF2S_0ywbliNq76w_t2Nnzju_B-Z2Rw`

When such token is provided authorization backend responses (to log) with message:
```
2019-12-08 14:30:11.233 [warning] <0.874.0> 
HTTP access denied: Authentication using an OAuth 2/JWT token failed: 
{invalid_aud, {resource_id_not_found_in_aud,<<"rabbitmq">>, [<<"http://localhost:50000/resources">>]}}
```

How to tests it? 
I prepared dedicated solution with preconfigured IdentityServer4. README.md file contains manual how to setup and use it.

Case where we do need use any ComplexClaims:
(please note that any user will have the same perrmision set)

advanced.config
``` 
{rabbitmq_management, [
	{enable_uaa, true},
	{uaa_client_id, "GitHub.RabbitMq.ManagementWebsite.Local"},
	{uaa_location, "http://localhost:50000"},
	{oauth2_scopes, "openid rabbitmq.configure:*/* rabbitmq.read:*/* rabbitmq.write:*/* rabbitmq.tag:monitoring"},
	{oauth2_implementation, identityserver}
  ]},
  
{rabbitmq_auth_backend_oauth2, [
	{resource_server_id, <<"rabbitmq">>},
	{key_config, [
		{signing_keys, #{
			<<"OXNwd_TLdbIcOex6bLLStQ">> => {map, #{	
				<<"kty">> => <<"RSA">>,
				<<"alg">> => <<"RS256">>,
				<<"e">> => <<"AQAB">>,
				<<"n">> => <<"sfq4sO8d2EPTxQqb56JTHNZ0OYr00qXfhxubmJUoiUTWjmTxtD48hTP2j-vgG4qw0eKeZyV8qZpA0q5w9HLk5xXNghwf5KMUDGiNhFd_5URBpVBcZ0fDSX2FwDTyRbF0XrxiYofROQeVOxbMIXHD0vCzTrzoypD6AA6wWI1ej4t-Zn5mpCG4DijIe-Ki0MJ4D58yY6YmuoERPhFNW1DmSKSohhex89anfrBfkIoyaBg9KNMtQW7oky5sPDH0GYqEWeXf4jqv4ODL-FtJhi74inUc2knXgznlkC15r-X_ubPv0ngvxetUELJFReWqyQDELwk_CBLZtxUiCkMZNl6JgQ">>
				}}
		}}
	]}
]}
```

Case where we use ComplexClaims:
(please note that each user may have his own permisions set included in _extra_scopes_source_ JWT section)

advanced.config
```
{rabbitmq_management, [
	{enable_uaa, true},
	{uaa_client_id, "GitHub.RabbitMq.ManagementWebsite.Local"},
	{uaa_location, "http://localhost:50000"},
	{oauth2_scopes, "openid rabbitmq.managementwebsite"},
	{oauth2_implementation, identityserver}
  ]},
  
{rabbitmq_auth_backend_oauth2, [
	{resource_server_id, <<"rabbitmq">>},
	{extra_scopes_source, <<"rabbit_resources">>},
	{key_config, [
		{signing_keys, #{
			<<"OXNwd_TLdbIcOex6bLLStQ">> => {map, #{	
				<<"kty">> => <<"RSA">>,
				<<"alg">> => <<"RS256">>,
				<<"e">> => <<"AQAB">>,
				<<"n">> => <<"sfq4sO8d2EPTxQqb56JTHNZ0OYr00qXfhxubmJUoiUTWjmTxtD48hTP2j-vgG4qw0eKeZyV8qZpA0q5w9HLk5xXNghwf5KMUDGiNhFd_5URBpVBcZ0fDSX2FwDTyRbF0XrxiYofROQeVOxbMIXHD0vCzTrzoypD6AA6wWI1ej4t-Zn5mpCG4DijIe-Ki0MJ4D58yY6YmuoERPhFNW1DmSKSohhex89anfrBfkIoyaBg9KNMtQW7oky5sPDH0GYqEWeXf4jqv4ODL-FtJhi74inUc2knXgznlkC15r-X_ubPv0ngvxetUELJFReWqyQDELwk_CBLZtxUiCkMZNl6JgQ">>
				}}
		}}
	]}
]}
```

 ## Types of Changes

What types of changes does your code introduce to this project?

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask on the mailing list. We're here to help! This is simply a reminder
of what we are going to look for before merging your code._

- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] Any dependent changes have been merged and published in related repositories
